Quark_Sketch/src/smc.cpp

```diff
@@ -0,0 +1,395 @@
+#endif
+
+#include "soc.h"
+#include "curie_shared_mem.h"
+
+#define SCSS_SS_CFG_REG (uint32_t*)0xb0800600
+#define SCSS_SS_STS_REG (uint32_t*)0xb0800604
+#define SCSS_RSTC   (uint32_t*)0xb0800570
+
+#define MMIO_REG_VAL_FROM_BASE(base, offset) (*((volatile uint32_t *)(base+offset)))
+#define SCSS_REG_VAL(offset) MMIO_REG_VAL_FROM_BASE(SCSS_REGISTER_BASE, offset)
+
+#define RSTC_WARM_RESET	(1 << 1)
+#define RSTC_COLD_RESET (1 << 3)
+
+#define SOFTRESET_INTERRUPT_PIN		0
+
+static volatile bool data_transmitted;
+static volatile bool data_arrived = false;
+
+
+struct device *dev;
+bool usbSetupDone = false;
+
+struct gpio_callback cb;
+
+
+
+/**
+ * Use the following defines just to make the tips of your finger happier.
+ */
+
+#define Rx_BUFF curie_shared_data->cdc_acm_shared_rx_buffer.data
+#define Rx_HEAD curie_shared_data->cdc_acm_shared_rx_buffer.head
+#define Rx_TAIL curie_shared_data->cdc_acm_shared_rx_buffer.tail
+#define Tx_BUFF curie_shared_data->cdc_acm_shared_tx_buffer.data
+#define Tx_HEAD curie_shared_data->cdc_acm_shared_tx_buffer.head
+#define Tx_TAIL curie_shared_data->cdc_acm_shared_tx_buffer.tail
+#define SBS     SERIAL_BUFFER_SIZE
+
+/* Make sure BUFFER_LENGTH is not bigger then shared ring buffers */
+#define BUFFER_LENGTH		128
+
+#define LOOP_INTERVAL_MS 1
+#define USB_ACM_TIMEOUT_MS (250)
+
+#define USB_CONNECTED	    0x04
+#define USB_DISCONNECTED    0x05
+
+// buffers
+static unsigned char data_buf[128];
+static uint8_t write_buffer[BUFFER_LENGTH*2];
+
+typedef enum {
+	ACM_RX_DISABLED,
+	ACM_RX_READY,
+	ACM_RX_PENDING
+} acm_rx_states;
+
+typedef enum {
+	ACM_TX_DISABLED,
+	ACM_TX_READY,
+	ACM_TX_PENDING
+} acm_tx_states;
+
+static volatile uint32_t acm_rx_state = ACM_RX_DISABLED;
+static volatile uint32_t acm_tx_state = ACM_TX_DISABLED;
+
+void cdc_acm_tx();
+void cdc_acm_rx();
+void softResetButton();
+void reboot();
+
+static void softReset_button_callback(struct device *port, struct gpio_callback *cb, uint32_t pins)
+{
+	PRINT("SOFT RESET\r\n");
+	reboot();
+}
+
+static void interrupt_handler(struct device *dev)
+{
+	uart_irq_update(dev);
+
+	if (uart_irq_tx_ready(dev)) {
+		data_transmitted = true;
+	}
+
+	if (uart_irq_rx_ready(dev)) {
+		data_arrived = true;
+	}
+}
+
+static void read_data(struct device *dev, int *bytes_read)
+{
+	int timeout = 1000;
+
+	while (data_arrived == false)
+	{
+		if(!timeout)
+			break;
+		timeout--;
+	}
+
+	data_arrived = false;
+
+	/* Read all data */
+	*bytes_read = uart_fifo_read(dev, data_buf, sizeof(data_buf));
+}
+
+static void write_data(struct device *dev, const char *buf, int len)
+{
+	uart_irq_tx_enable(dev);
+
+	data_transmitted = false;
+	uart_fifo_fill(dev, (const uint8_t*)buf, len);
+	while (data_transmitted == false)
+		;
+
+	uart_irq_tx_disable(dev);
+}
+
+void start_arc(unsigned int reset_vector)
+{
+
+	if (reset_vector != 0) {
+		curie_shared_data->arc_start = reset_vector;
+	}
+
+	curie_shared_data->flags = 0;
+	for(int i = 0; i < 6400000; i++)
```

Comment:
The compiler will not generate code for this loop.  It has optimization setting on (highest level too, if my memory serves me correctly).